buffer/src/main/java/io/netty/buffer/api/BufferHolder.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import io.netty.buffer.api.internal.ResourceSupport;
+import io.netty.buffer.api.internal.Statics;
+
+import java.lang.invoke.VarHandle;
+import java.util.Objects;
+
+import static java.lang.invoke.MethodHandles.lookup;
+
+/**
+ * The {@link BufferHolder} is an abstract class that simplifies the implementation of objects that themselves contain
+ * a {@link Buffer} instance.
+ * <p>
+ * The {@link BufferHolder} can only hold on to a single buffer, so objects and classes that need to hold on to multiple
+ * buffers will have to do their implementation from scratch, though they can use the code of the {@link BufferHolder}
+ * as inspiration. Alternatively, multiple buffers can be
+ * {@linkplain CompositeBuffer#compose(BufferAllocator, Send[]) composed} into a single buffer, which can then be put
+ * in a buffer holder.
+ * <p>
+ * If you just want an object that is a reference to a buffer, then the {@link BufferRef} can be used for that purpose.
+ * If you have an advanced use case where you wish to implement {@link Resource}, and tightly control lifetimes, then
+ * {@link ResourceSupport} can be of help.
+ *
+ * @param <T> The concrete {@link BufferHolder} type.
+ */
+public abstract class BufferHolder<T extends BufferHolder<T>> implements Resource<T> {
+    private static final VarHandle BUF = Statics.findVarHandle(lookup(), BufferHolder.class, "buf", Buffer.class);
+    private Buffer buf;
+
+    /**
+     * Create a new {@link BufferHolder} to hold the given {@linkplain Buffer buffer}.
+     *
+     * @param buf The {@linkplain Buffer buffer} to be held by this holder.
+     */
+    protected BufferHolder(Buffer buf) {
+        this.buf = Objects.requireNonNull(buf, "The buffer cannot be null.");
+    }
+
+    /**
+     * Create a new {@link BufferHolder} to hold the {@linkplain Buffer buffer} received from the given {@link Send}.
+     * <p>
+     * The {@link BufferHolder} will then be holding exclusive ownership of the buffer.
+     *
+     * @param send The {@linkplain Buffer buffer} to be held by this holder.
+     */
+    protected BufferHolder(Send<Buffer> send) {
```

Comment:
`Send<? extends Buffer>`

Reply:
This might go along with making `CompositeBuffer` and interface, so I might do that in a different PR.